src/librbd/io/SimpleSchedulerObjectDispatch.cc

```diff
@@ -0,0 +1,516 @@
+    const ::SnapContext &snapc, int op_flags,
+    const ZTracer::Trace &parent_trace, uint64_t* mismatch_offset,
+    int* object_dispatch_flags, uint64_t* journal_tid,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << cmp_data.length() << dendl;
+
+  Mutex::Locker locker(m_lock);
+  dispatch_delayed_requests(object_no, on_finish);
+
+  return false;
+}
+
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::flush(
+    io::FlushSource flush_source, const ZTracer::Trace &parent_trace,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  dispatch_all_delayed_requests(on_finish);
+
+  return false;
+}
+
+template <typename I>
+bool SimpleSchedulerObjectDispatch<I>::try_delay_write(
+    uint64_t object_no, uint64_t object_off, ceph::bufferlist&& data,
+    const ::SnapContext &snapc, int op_flags, int object_dispatch_flags,
+    Context* on_dispatched) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+
+  if (m_latency_stats && !m_latency_stats->is_ready()) {
+    ldout(cct, 20) << "latency stats not collected yet" << dendl;
+    return false;
+  }
+
+  auto it = m_requests.find(object_no);
+  if (it == m_requests.end()) {
+    ldout(cct, 20) << "no pending requests" << dendl;
+    return false;
+  }
+
+  auto &object_requests = it->second;
+  bool delayed = object_requests->try_delay_request(
+      object_off, std::move(data), snapc, op_flags, object_dispatch_flags,
+      on_dispatched);
+
+  ldout(cct, 20) << "delayed: " << delayed << dendl;
+
+  // schedule dispatch on the first request added
+  if (delayed && !object_requests->is_scheduled_dispatch()) {
+    auto dispatch_time = ceph_clock_now();
+    if (m_latency_stats) {
+      dispatch_time += utime_t(0, m_latency_stats->avg() / 2);
+    } else {
+      dispatch_time += utime_t(0, m_max_delay * 1000000);
+    }
+    object_requests->set_scheduled_dispatch(dispatch_time);
+    m_dispatch_queue.push_back(object_requests);
+    if (m_dispatch_queue.front() == object_requests) {
+      schedule_dispatch_delayed_requests();
+    }
+  }
+
+  return delayed;
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_all_delayed_requests(
+    Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << dendl;
+
+  for (auto &it : m_requests) {
+    dispatch_delayed_requests(it.first, on_finish);
+  }
+}
+
+template <typename I>
+void SimpleSchedulerObjectDispatch<I>::dispatch_delayed_requests(
+    uint64_t object_no, Context** on_finish) {
+  ceph_assert(m_lock.is_locked());
+
+  dispatch_delayed_requests(object_no);
+
+  auto res = m_requests.insert(
```

Comment:
Is this causing reads (and other ops) to insert a new entry in the in-flight requests list? Should only measure latency for writes.

Reply:
Yes, right now:
1) a write is delayed if there is any in-flight op;
2) latency is measured for any request.

I see you want (2) to be changed to measure latency only for writes. What is about (1)? Would you prefer a write is delayed only if there is a pending write (non read) request?

Reply:
Just my opinion, but I don't think there should be any reason to delay a write if a read in in-flight.